compiler/next/include/chpl/queries/stringify-functions.h

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This file implements the generic chpl::stringify
+  as well as specializations for some common types.
+
+  stringify can be used to get a string representation of an object
+  for debugging purposes
+ */
+
+#ifndef CHPL_QUERIES_STRINGIFY_FUNCTIONS_H
+#define CHPL_QUERIES_STRINGIFY_FUNCTIONS_H
+
+#include <cassert>
+#include <cstring>
+#include <functional>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "chpl/util/memory.h"
+#include <set>
+#include <sstream>
+
+
+namespace chpl {
+class Context;
+
+enum StringifyKind {
+  DEBUG_SUMMARY,
+  DEBUG_DETAIL,
+  CHPL_SYNTAX
+};
+
+
+template<typename T> struct stringify {
+  std::string operator()(StringifyKind stringKind, const T& stringMe) const = 0;
+};
+
+
+template<typename T> struct stringify<const T*> {
+  std::string operator()(StringifyKind stringKind, const T* stringMe) const {
+    stringify<T> stringifier;
+    return stringifier(stringKind, *stringMe);
+  }
+};
+
+template<typename T>
+static inline std::string defaultStringify(StringifyKind stringKind,
+                                           const T& stringMe) {
+
+  return "HAVE NOT IMPLEMENTED STRINGIFY YET";
+
+}
+
+template<typename T>
+static inline std::string defaultStringifyVec(StringifyKind stringKind,
+                                              const std::vector<T>& stringVec)
+{
+  if (stringVec.empty()) {
+   return "[ ]";
+  } else {
+   std::ostringstream ss;
+   std::string separator;
+   stringify<T> vecString;
+   for (const auto &vecVal : stringVec ) {
+     ss << separator << vecString(stringKind, vecVal);
+     separator = ",";
+   }
+   return std::string("["+ss.str()+"]");
+  }
+}
+
+template<typename K, typename V>
+static inline std::string defaultStringifyMap(StringifyKind stringKind,
+                                              const std::unordered_map<K,V>& stringMap)
+{
+  if (stringMap.size() == 0) {
+    return "{ }";
+  } else {
+    std::ostringstream ss;
+    std::string separator;
+    stringify<K> keyString;
+    stringify<V> valString;
+    for (auto const& x : stringMap)
```

Comment:
since it's an unordered map this iteration will occur in a nondeterministic order. it's important to sort the keys / iterate in a deterministic order here